File: 2025-02-February-GFG-POTD/February 19 - Merge K sorted linked lists/Solution.py
import heapq

class Node:
    def __init__(self, x):
        self.data = x
        self.next = None

class Solution:
    def mergeKLists(self, arr):
        minHeap = []

        # Add all the head nodes to the heap
        for head in arr:
            if head:
                heapq.heappush(minHeap, (head.data, head))

        # Dummy node for merged list
        dummy = Node(0)
        current = dummy

        while minHeap:
            _, smallestNode = heapq.heappop(minHeap)

            current.next = smallestNode
            current = current.next

            if smallestNode.next:
                heapq.heappush(minHeap, (smallestNode.next.data, smallestNode.next))

        return dummy.next

# Function to print the linked list
def printList(node):
    while node:
        print(node.data, end=" ")
        node = node.next
    print()

# Driver Code
if __name__ == "__main__":
    t = int(input())

    for _ in range(t):
        n = int(input())
        lists = []

        for _ in range(n):
            values = list(map(int, input().split()))
            head = None
            tail = None

            for val in values:
                newNode = Node(val)
                if head is None:
                    head = newNode
                    tail = head
                else:
                    tail.next = newNode
                    tail = tail.next

            lists.append(head)

        sol = Solution()
        result = sol.mergeKLists(lists)
        printList(result)
        print("~")

File: 2025-02-February-GFG-POTD/February 19 - Merge K sorted linked lists/Solution.cpp
#include <bits/stdc++.h>
using namespace std;

class Node {
public:
    int data;
    Node* next;

    Node(int x) {
        data = x;
        next = nullptr;
    }
};

// Comparison function for the priority queue (min-heap)
struct Compare {
    bool operator()(Node* a, Node* b) {
        return a->data > b->data;
    }
};

class Solution {
public:
    Node* mergeKLists(vector<Node*>& arr) {
        priority_queue<Node*, vector<Node*>, Compare> minHeap;

        // Add heads of all linked lists to the min-heap
        for (Node* head : arr) {
            if (head) {
                minHeap.push(head);
            }
        }

        // Dummy node for easier list construction
        Node* dummy = new Node(0);
        Node* current = dummy;

        while (!minHeap.empty()) {
            Node* smallestNode = minHeap.top();
            minHeap.pop();

            current->next = smallestNode;
            current = current->next;

            if (smallestNode->next) {
                minHeap.push(smallestNode->next);
            }
        }

        return dummy->next;
    }
};

// Function to print the linked list
void printList(Node* node) {
    while (node != nullptr) {
        cout << node->data << " ";
        node = node->next;
    }
    cout << endl;
}

// Driver code
int main() {
    int t;
    cin >> t;

    while (t--) {
        int n;
        cin >> n;
        vector<Node*> lists(n);

        for (int i = 0; i < n; i++) {
            int val;
            Node* head = nullptr;
            Node* tail = nullptr;
            while (cin >> val) {
                if (val == -1) break; // Assuming -1 as the end of a linked list
                Node* newNode = new Node(val);
                if (!head) {
                    head = newNode;
                    tail = head;
                } else {
                    tail->next = newNode;
                    tail = tail->next;
                }
            }
            lists[i] = head;
        }

        Solution sol;
        Node* result = sol.mergeKLists(lists);
        printList(result);
        cout << "~" << endl;
    }

    return 0;
}
